#6 - 상품 등록 API 구현 및 각 레이어별 테스트 코드 작성

ChaYujin opened the pull request:
- POST /products (MediaType : multipart/form-data)로 상품 등록하는 API 구현
- 컨트롤러, 서비스, 레파지토리 레이어의 단위테스트 코드 작성

This closes #6

ChaYujin committed:
#6 - 상품 등록 API 구현

- 이미지 정보와 함께 받아 처리하기위해 POST /products (mediaType: multipart/form-data)  요청을 받으면 상품 등록 작업을 처리
- ProductService에서 상품 등록에 대한 비즈니스 로직을 처리함. (유효성검사, 엔티티 변환, 엔티티 저장, 응답값 생성)
- 마켓정보와 이미지정보가 잘못됐을 때 예외를 처리하기위한 Exception들 추가
- 반복되는 응답값 생성 코드를 줄이기위해 ApiResponse 클래스 추가

ChaYujin committed:
#6 - 상품 등록의 정상 케이스에 대한 단위테스트 코드 작성

ProductController 단위테스트 코드로 상품 등록에 대한 정상 케이스의 테스트코드를 작성함

ChaYujin committed:
#6 - Configurer에 ProductStatusCodeRequestConverter 등록하는 코드 추가

상품 등록 API에서 파라미터로 넘어오는 ProductStatusCode를 enum값으로 변환하는 작업이 필요하다. 이를 위해 ProductStatusCodeRequestConverter를 WebMvcConfigurer에 등록함

ChaYujin committed:
#6 - 예외발생시 ErrorResponseDto를 ResponseBody로 보내도록 수정

문자열 json을 직접 만들지 않고 DTO를 반환하여 자동으로 변환되어 ResponseBody 정보로 내보내도록 수정

ChaYujin committed:
#6 - 기존에 String을 반환하는 ExceptionHandler를 ErrorResponseDto로 반환하도록 수정

기존에 있던 ExceptionHandler에서 json 문자열을 직접 만들어 반환하도록 하였다. 이렇게 반환하게 되면 content-type이 plain/text로 응답이 되게 된다. 응답값의 content-type이 applcation/json으로 넘기기위해 ErrorResponseDto를 반환하여 스프링에 등록된 컨버터에 의해 json으로 컨버팅이 되도록 수정하였다.

ChaYujin committed:
#6 - 상품 등록에 대한 성공케이스, 실패케이스 테스트코드 추가

- 이미지 정보가 없어도 상품 등록이 되는 성공 케이스 테스트코드 작성
- 필수 파라미터 누락으로 Exception이 발생하여 상품 등록에 실패하는 케이스 에 대한 테스트코드 작성

ChaYujin committed:
#6 - MultipartFile객체에 대한 방어코드 추가, catch하는 Exception을 구체적인 Exception으로 변경

image값은 필수값이 아니기때문에 값이 없을 수도 있다. 그 경우를 대비하여 방어코드를 추가함

ChaYujin committed:
#6 - Product Repository 단위테스트 코드 작성

JPA레이어에 대한 단위테스트를 진행하는 테스트코드 작성
select, insert, update, delete에 대한 성공 케이스에 대한 테스트코드 작성

ChaYujin committed:
#6 - Product엔티티 필드값 업데이트하는 메소드 추가

수정될 가능성이 있는 필드들을 수정할 수 있는 메소드 추가

ChaYujin committed:
#6 - Product Service 단위테스트 코드 작성

Product 생성에 대한 서비스 레이어의 단위테스트 코드를 작성함
- 생성 성공, 실패 케이스에 대한 단위테스트 작성

chanwoothecali commented on src/main/java/com/flab/modu/global/exception/GlobalExceptionHandler.java:
```diff
@@ -10,19 +14,27 @@
-                + "\"}"
-        );
+    public ResponseEntity<Object> errorHandler(MethodArgumentNotValidException e) {
+        return ApiResponse.buildResponseEntity(e.getAllErrors().get(0).getDefaultMessage());
```
ErrorResponseDto가 String만을 담고있는 객체라면 ResponseStatusException로 반환하는 게 어떨까 싶습니다!

f-lab-michael commented on src/main/java/com/flab/modu/market/exception/MarketNotFoundException.java:
```diff
@@ -0,0 +1,5 @@
+package com.flab.modu.market.exception;
+
+public class MarketNotFoundException extends RuntimeException {
```
ResponseStatusException을 상속받아서 HttpStatus 를 설정하면 해당 status 로 응답이 가게 할수있습니다

f-lab-michael commented on src/main/java/com/flab/modu/product/domain/entity/Product.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Lob
+    @Column(length = 10240)
+    private byte[] image;
```
일반적으로 이미지는 디비 직접 저장보다는 별도의 스토리지 서비스에 업로드후 해당 이미지 url 정보를 디비에 넣는 방식을 사용합니다.

f-lab-michael commented on src/main/java/com/flab/modu/product/service/ProductService.java:
```diff
@@ -0,0 +1,63 @@
+
+    public ProductDto.CreateResponse createProduct(ProductDto.CreateRequest createRequest,
+        MultipartFile imageMultipartFile, String loginId) {
+        //유효성 검사
```
이런경우 comment 로 설명을 하는것보다 해당 부분의 코드를 별도의 메소드로 분리해서 이름을 가독성있게 지어주는것이 더 좋습니다.
아래에 존재하는 코드상의 comment 모두 동일합니다.

ChaYujin committed:
#6 - 코드리뷰 피드백 반영

- 예측된 예외를 처리할 때, 매번 ResponseEntity객체를 만들어 반환하지 않고 발생시킬 Exception에 ResponseStatusException을 상속받아 원하는 HttpStatus값과 에러 메시지를 보내도록 처리
- ProductService의 createProduct메서드안에서 처리하는 내용을 좀 더 직관적으로 알 수 있도록 의미있는 메소드로 분리
- Controller레이어 단위테스트에 service에서 발생시킬 수 있는 Exception처리에 대한 테스트코드 작성

ChaYujin committed:
#6 - 필수값이 아닌 파라메터에 required=false 추가 및 테스트코드 변경

필수값이 아닌 파라메터에 대해 required=false로 처리하는게 맞다고 판단하여 required속성값을 추가하고, 이와 관련된 테스트코드에 mockMvc로 요청할 때 해당 파라메터 자체를 넘기지 않도록 수정함

f-lab-michael commented:
@ChaYujin 컨플릭 수정해주세요~

ChaYujin committed:
Merge branch 'develop' into feature/6/product

ChaYujin commented:
> @ChaYujin 컨플릭 수정해주세요~

@f-lab-michael 컨플릭 수정했습니다 :)
